server/clustering_test.go

```diff
@@ -0,0 +1,323 @@
+	deadline := time.Now().Add(timeout)
+	for time.Now().Before(deadline) {
+		for i := 0; i < len(servers); i++ {
+			s := servers[i]
+			if s.state == Shutdown {
+				continue
+			}
+			s.channels.Lock()
+			c := s.channels.channels[channel]
+			s.channels.Unlock()
+			if c == nil || c.raft == nil {
+				continue
+			}
+			if c.isLeader() {
+				if leader != nil {
+					stackFatalf(t, "Found more than one channel leader")
+				}
+				leader = s
+			}
+		}
+		if leader != nil {
+			break
+		}
+	}
+	if leader == nil {
+		stackFatalf(t, "Unable to find the channel leader")
+	}
+	return
+}
+
+func removeServer(servers []*StanServer, s *StanServer) []*StanServer {
+	for i, srv := range servers {
+		if srv == s {
+			servers = append(servers[:i], servers[i+1:]...)
+		}
+	}
+	return servers
+}
+
+func assertMsg(t *testing.T, msg pb.MsgProto, expectedData []byte, expectedSeq uint64) {
+	if !bytes.Equal(msg.Data, expectedData) {
+		stackFatalf(t, "Msg data incorrect, expected: %s, got: %s", expectedData, msg.Data)
+	}
+	if msg.Sequence != expectedSeq {
+		stackFatalf(t, "Msg sequence incorrect, expected: %d, got: %d", expectedSeq, msg.Sequence)
+	}
+}
+
+func TestClusteringConfig(t *testing.T) {
+	cleanupDatastore(t)
+	defer cleanupDatastore(t)
+	cleanupRaftLog(t)
+	defer cleanupRaftLog(t)
+
+	opts := GetDefaultOptions()
+	opts.ClusterPeers = []string{"a", "b"}
+	s, err := RunServerWithOpts(opts, nil)
+	if s != nil || err == nil {
+		s.Shutdown()
+		t.Fatal("Server should have failed to start with cluster peers and no cluster node id")
+	}
+}
+
+// Ensure basic replication works as expected. This test starts three servers
+// in a cluster, publishes messages to the cluster, kills the leader, publishes
+// more messages, kills the new leader, verifies progress cannot be made when
+// there is no leader, then brings the cluster back online and verifies
+// catchup and consistency.
+func TestClusteringBasic(t *testing.T) {
+	cleanupDatastore(t)
+	defer cleanupDatastore(t)
+	cleanupRaftLog(t)
+	defer cleanupRaftLog(t)
+
+	// For this test, use a central NATS server.
+	ns := natsdTest.RunDefaultServer()
+	defer ns.Shutdown()
+
+	// Configure first server
+	s1sOpts := getTestDefaultOptsForClustering("a", []string{"b", "c"})
+	s1, err := RunServerWithOpts(s1sOpts, nil)
+	if err != nil {
+		t.Fatalf("Error starting server: %v", err)
+	}
+	defer s1.Shutdown()
+
+	// Configure second server.
+	s2sOpts := getTestDefaultOptsForClustering("b", []string{"a", "c"})
+	s2, err := RunServerWithOpts(s2sOpts, nil)
+	if err != nil {
+		t.Fatalf("Error starting server: %v", err)
+	}
+	defer s2.Shutdown()
+
+	// Configure third server.
+	s3sOpts := getTestDefaultOptsForClustering("c", []string{"a", "b"})
+	s3, err := RunServerWithOpts(s3sOpts, nil)
+	if err != nil {
+		t.Fatalf("Error starting server: %v", err)
+	}
+	defer s3.Shutdown()
+
+	servers := []*StanServer{s1, s2, s3}
+
```

Comment:
I would iterate this array and call: `checkState(t, s, Clustered)`. This will check that the state is properly set.